stl/src/locale0.cpp

```diff
@@ -135,8 +135,39 @@ __PURE_APPDOMAIN_GLOBAL locale::id ctype<unsigned short>::id(0);
 
 __PURE_APPDOMAIN_GLOBAL locale::id codecvt<unsigned short, char, mbstate_t>::id(0);
 
+#define _Compiler_barrier() _STL_DISABLE_DEPRECATED_WARNING _ReadWriteBarrier() _STL_RESTORE_DEPRECATED_WARNING
+
+#if defined(_M_ARM) || defined(_M_ARM64) || defined(_M_ARM64EC)
+#define _Memory_barrier()             __dmb(0xB) // inner shared data memory barrier
+#define _Compiler_or_memory_barrier() _Memory_barrier()
+#elif defined(_M_IX86) || defined(_M_X64)
+// x86/x64 hardware only emits memory barriers inside _Interlocked intrinsics
+#define _Compiler_or_memory_barrier() _Compiler_barrier()
+#else // ^^^ x86/x64 / unsupported hardware vvv
+#error Unsupported hardware
+#endif // hardware
+
 _MRTIMP2_PURE const locale& __CLRCALL_PURE_OR_CDECL locale::classic() { // get reference to "C" locale
-    _Init();
+#if !defined(_M_CEE_PURE)
+    const auto mem = reinterpret_cast<const intptr_t*>(&locale::_Locimp::_Clocptr);
+    intptr_t as_bytes;
+#ifdef _WIN64
+#ifdef _M_ARM
+    as_bytes = __ldrexd(_Mem);
+#else // ^^^ ARM64 / x64 vvv

```

Comment:
This is incorrect. `_M_ARM` indicates 32-bit ARM, and is never defined for ARM64. Example:

```
C:\Temp>"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvarsall.bat" x64_arm64
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.4.0-pre.4.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x64_arm64'

C:\Temp>cl
Microsoft (R) C/C++ Optimizing Compiler Version 19.34.31931 for ARM64
Copyright (C) Microsoft Corporation.  All rights reserved.

usage: cl [ option... ] filename... [ /link linkoption... ]

C:\Temp>type meow.cpp
```
```cpp
#if defined(_WIN64) && defined(_M_ARM)
static_assert(false, "defined(_WIN64) && defined(_M_ARM) was true");
#endif

#if defined(_WIN64) && defined(_M_ARM64)
static_assert(false, "defined(_WIN64) && defined(_M_ARM64) was true");
#endif
```
```
C:\Temp>cl /EHsc /nologo /W4 /c meow.cpp
meow.cpp
meow.cpp(6): error C2338: static_assert failed: 'defined(_WIN64) && defined(_M_ARM64) was true'
```

Reply:
I took it from the code in `_Atomic_storage<_Ty, 8>`:
https://github.com/microsoft/STL/blob/fb386c01083245a80db2666433235918c40d8363/stl/inc/atomic#L1027-L1036

It looks like we should unconditionally use `__iso_volatile_load64` for 64-bit, then.